### DISCUSSION TITLE
Virtualization Support

### DISCUSSION THREAD

**Discussion by SaavanNanavati** [ariakit/ariakit#1930] [Questions] [answered]
I want to use react-window with this library. Essentially tree structure would be:

<Combobox/>
   <ComboboxPopover/>
       <VirtualizationList/>
            <ComboboxItem/>
            <ComboboxItem/>
            <ComboboxItem/>
...

Would this work with keyboard nav still?

**Comment by diegohaz** [comment, marked as answer]
react-window doesn't work very well with keyboard navigation as far as I know, though I've seen people successfully implementing it with Ariakit composite components. But that's complicated.

We're working on a built-in virtualization feature, but this will take some time.

**Reply by SaavanNanavati** [reply]
Can you point me to an example where it's worked?

**Reply by diegohaz** [reply]
Unfortunately, I don't have access to the code.

**Reply by ericwilhite** [reply]
> We're working on a built-in virtualization feature, but this will take some time.

I saw in newsletter a note about some [virtualization features already existing](https://newsletter.ariakit.org/p/virtualization) via the Composite component. If this is true, I think it would be great to have some examples with Combobox. 

We have 1000s of items to render and would love to migrate from a custom HeadlessUI + Tanstack Virtual component that we cobbled together.

**Reply by diegohaz** [reply]
@ericwilhite There's an example here: https://ariakit.org/examples/select-combobox-virtualized

The `*Renderer` components are still experimental, though.
